### PR TITLE
Fix move semantics

### PR DESCRIPTION
**Pull request by Loki-Astari**
Fixed a potential bug with `UserType` where user supplied data could be deleted in one object when it is still potentially being used by another object (because the original object was copied).

Fixed a couple of other issues with move semantics (to make it work correclty)

### CODE REVIEW

**Comment by Loki-Astari** [comment]
This needs https://github.com/ArkScript-lang/modules/pull/46 to compile.
To show it compiles this commit has updated the modules to https://github.com/Loki-Astari/modules
I can revert this change once the above modules pull request has been merged.

**Comment by Loki-Astari** [comment]
I have a feeling this will fix your build issues on the Mac.
More than likely this fix UB on the other platforms (that is causing Mac to crash).

File: include/Ark/VM/UserType.hpp [hunk 1]
```diff
@@ -138,11 +137,10 @@ namespace Ark
 
     private:
         uint16_t m_type_id;
-        void* m_data;
-        ControlFuncs* m_funcs;
+        std::shared_ptr<void> m_data;
+        ControlFuncs* m_funcs;  // Not none owned pointer.
```
Comment (SuperFola) [inline include/Ark/VM/UserType.hpp:141]: `Not none` -> `non`

File: include/Ark/VM/VM.hpp [hunk 2]
```diff
@@ -119,7 +119,9 @@ namespace Ark
         friend class Repl;
 
     private:
-        State* m_state;
+        State* m_state;  // Note: This is a non owned pointer.
+                         //       This should probably be changed to be a reference to
+                         //       show that the object does not take ownership.
```
Comment (SuperFola) [inline include/Ark/VM/VM.hpp:124]: Might as well slap a todo on this

File: include/Ark/VM/UserType.hpp [hunk 3]
```diff
@@ -65,32 +66,30 @@ namespace Ark
...
-         */
-        void del();
-
         /**
          * @brief Set the control functions structure
          * 
          * @param block A pointer to an instance of this block
          */
-        inline void setControlFuncs(ControlFuncs* block) noexcept;
+        // inline void setControlFuncs(ControlFuncs* block) noexcept;
```
Comment (SuperFola) [inline include/Ark/VM/UserType.hpp:82]: Might as well delete it

**Review by SuperFola — APPROVED** [review]
LGTM

**Pull request merged**